iterator/cdc.go

```diff
@@ -0,0 +1,265 @@
+		p:                  p,
+	}
+	shardIterator, err := c.getShardIterator(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get shard iterator: %w", err)
+	}
+	c.shardIterator = shardIterator
+
+	// start listening to changes
+	c.tomb.Go(c.startCDC)
+
+	return c, nil
+}
+
+// HasNext returns a boolean that indicates whether the iterator has any objects in the buffer or not.
+func (c *CDCIterator) HasNext(_ context.Context) bool {
+	return len(c.cache) > 0 || !c.tomb.Alive() // if tomb is dead we return true so caller will fetch error with Next
+}
+
+// Next returns the next record in the iterator.
+func (c *CDCIterator) Next(ctx context.Context) (opencdc.Record, error) {
+	var rec stypes.Record
+	select {
+	case r := <-c.cache:
+		rec = r
+	case <-c.tomb.Dead():
+		return opencdc.Record{}, fmt.Errorf("tomb is dead: %w", c.tomb.Err())
+	case <-ctx.Done():
+		return opencdc.Record{}, ctx.Err()
+	}
+
+	return c.getOpenCDCRec(rec)
+}
+
+func (c *CDCIterator) Stop() {
+	// stop the goRoutine
+	c.ticker.Stop()
+	c.tomb.Kill(errors.New("cdc iterator is stopped"))
+}
+
+// startCDC gets records from the stream using the shardIterators,
+// then adds the records to a buffered cache.
+func (c *CDCIterator) startCDC() error {
+	defer close(c.cache)
+
+	for {
+		select {
+		case <-c.tomb.Dying():
+			return fmt.Errorf("tomb is dying: %w", c.tomb.Err())
+		case <-c.ticker.C: // detect changes every polling period
+			out, err := c.streamsClient.GetRecords(c.tomb.Context(nil), &dynamodbstreams.GetRecordsInput{ //nolint:staticcheck // SA1012 tomb expects nil
+				ShardIterator: c.shardIterator,
+			})
+			if err != nil {
+				return fmt.Errorf("error getting records: %w", err)
+			}
+
+			// process the stream records
+			for _, record := range out.Records {
+				select {
+				case c.cache <- record:
+					// Successfully sent record
+				case <-c.tomb.Dying():
+					return fmt.Errorf("tomb is dying: %w", c.tomb.Err())
+				}
+			}
+
+			// update the shard iterator
+			c.shardIterator = out.NextShardIterator
+
+			if c.shardIterator == nil {
+				c.shardIterator, err = c.getShardIterator(c.tomb.Context(nil)) //nolint:staticcheck // SA1012 tomb expects nil
+				if err != nil {
+					return fmt.Errorf("failed to get shard iterator: %w", err)
+				}
+			}
+		}
+	}
+}
+
+// getShardIterator gets the shard iterator for the stream.
+func (c *CDCIterator) getShardIterator(ctx context.Context) (*string, error) {
+	// Describe the stream to get the shard ID
+	describeStreamOutput, err := c.streamsClient.DescribeStream(ctx, &dynamodbstreams.DescribeStreamInput{
+		StreamArn: aws.String(c.streamArn),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to describe stream: %w", err)
+	}
+
+	// Start from the last shard
+	if c.shardIndex == 0 {
+		c.shardIndex = len(describeStreamOutput.StreamDescription.Shards) - 1
+	} else {
+		c.shardIndex++
+	}
+	shardID := describeStreamOutput.StreamDescription.Shards[c.shardIndex].ShardId
```

Comment:
Here's how I understood, let me know if that's correct: the idea is to get changes from each shard. When we're done with changes from one shard, we go to the next one. The first time a CDC iterator is created (e.g. when we switch from snapshot to CDC, or when a connector that was already in CDC is restarted), we start with the last shard.

Reply:
the case of restarting while in CDC is still not handled, for that I'm gonna have to start reading a shard from a specific "sequence number", it's still a todo.
But for changing from snapshot to CDC, I start reading from the last shard because any new data will be added there, old shards will have old events that were done before the connector was on, so we don't need those.
We start reading from the last shard, when this shard starts returning nil, then it has expired and a newer shard is created, so we start reading from the next shard, which will have the (last shards index)+1.

Reply:
the update event shows both the old record as whole, and the new record.

> old shards will have old events that were done before the connector was on, so we don't need those.

what I meant here was that the user could have had the Stream running before they started conduit, so it would have some events that are already there in the table and will be picked up by the snapshot iterator, so I wanted the CDC iterator to skip the old data that's there before conduit even ran. and this is done by starting listening to the most recent shard, cuz the older shards would have old data that we don't want.